src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -1724,109 +1870,505 @@ ep_rt_mono_write_event_method_load (
+{
+	if (!EventEnabledExceptionThrown_V1 ())
+		return true;
+
+	if (obj) {
+		ERROR_DECL (error);
+		char *type_name = NULL;
+		char *exception_message = NULL;
+		uint16_t flags = 0;
+		uint32_t hresult = 0;
+		uintptr_t ip = 0;
+
+		if (mono_object_isinst_checked ((MonoObject *) obj, mono_get_exception_class (), error)) {
+			MonoException *exception = (MonoException *)obj;
+			flags |= EXCEPTION_THROWN_FLAGS_IS_CLS_COMPLIANT;
+			if (exception->inner_ex)
+				flags |= EXCEPTION_THROWN_FLAGS_HAS_INNER;
+			exception_message = ep_rt_utf16_to_utf8_string (mono_string_chars_internal (exception->message), mono_string_length_internal (exception->message));
+			hresult = exception->hresult;
+		}
+
+		if (mono_get_eh_callbacks ()->mono_walk_stack_with_ctx)
+			mono_get_eh_callbacks ()->mono_walk_stack_with_ctx (get_exception_ip_func, NULL, MONO_UNWIND_SIGNAL_SAFE, (void *)&ip);
+
+		type_name = mono_type_get_name_full (m_class_get_byval_arg (mono_object_class (obj)), MONO_TYPE_NAME_FORMAT_IL);
+
+		FireEtwExceptionThrown_V1 (
+			type_name,
+			exception_message,
+			(void *)&ip,
+			hresult,
+			flags,
 			clr_instance_get_id (),
 			NULL,
 			NULL);
 
-		g_free (assembly_name);
+		g_free (exception_message);
+		g_free (type_name);
+	}
+
+	return true;
+}
+
+bool
+ep_rt_mono_write_event_exception_clause (
+	MonoMethod *method,
+	uint32_t clause_num,
+	MonoExceptionEnum clause_type,
+	MonoObject *obj)
+{
+	if ((clause_type == MONO_EXCEPTION_CLAUSE_FAULT || clause_type == MONO_EXCEPTION_CLAUSE_NONE) && (!EventEnabledExceptionCatchStart() || !EventEnabledExceptionCatchStop()))
+		return true;
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FILTER && (!EventEnabledExceptionFilterStart() || !EventEnabledExceptionFilterStop()))
+		return true;
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FINALLY && (!EventEnabledExceptionFinallyStart() || !EventEnabledExceptionFinallyStop()))
+		return true;
+
+	uintptr_t ip = 0; //TODO: Have profiler pass along IP of handler block.
+	uint64_t method_id = (uint64_t)method;
+	char *method_name = NULL;
+
+	method_name = mono_method_get_name_full (method, TRUE, TRUE, MONO_TYPE_NAME_FORMAT_IL);
+
+	if ((clause_type == MONO_EXCEPTION_CLAUSE_FAULT || clause_type == MONO_EXCEPTION_CLAUSE_NONE)) {
+		FireEtwExceptionCatchStart (
+			(uint64_t)ip,
+			method_id,
+			(const ep_char8_t *)method_name,
+			clr_instance_get_id (),
+			NULL,
+			NULL);
+
+		FireEtwExceptionCatchStop (
+			NULL,
+			NULL);
+	}
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FILTER) {
+		FireEtwExceptionFilterStart (
+			(uint64_t)ip,
+			method_id,
+			(const ep_char8_t *)method_name,
+			clr_instance_get_id (),
+			NULL,
+			NULL);
+
+		FireEtwExceptionFilterStop (
+			NULL,
+			NULL);
+	}
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FINALLY) {
+		FireEtwExceptionFinallyStart (
+			(uint64_t)ip,
+			method_id,
+			(const ep_char8_t *)method_name,
+			clr_instance_get_id (),
+			NULL,
+			NULL);
+
+		FireEtwExceptionFinallyStop (
```

Comment:
I'm curious what these Start/Stop events are actually meant to measure.  The duration of exception handling?  What if I have a method call inside a `finally` clause: should the start/stop events span the duration of the method call?  I wonder if these 0 duration events would be more misleading/confusing than not emitting them at all.

Reply:
Yes, this is one case where we don't have a good match, and right now they are not even emitted since it needs specific instrumentation that must be enabled at startup.

CoreCLR use it around its call to handler in ExceptionTracker::CallHandler.

Reply:
So the timing will be incorrect in our case, since we don't have corresponding profiler events, but I believe the events might still have a value since they give you much more information than just the time since they include method info, catch/filter/final as well as callstack info.

Reply:
Alternative could be to add similar exception handling start/stop to mono profiler.

Reply:
But it's not a problem if calls from inside the handler to other methods have timestamps that don't nest? (ie: we're in a method "A" from time 0 to time 2; we enter the finally clause at time 1, call method "Cleanup" from time 1.1 to 1.5, then leave the finally clause at time 1(!), then leave the method "A" at time 2)

Reply:
EventPipe doesn't currently have events for full method tracing (enter/leave etc), so we only have specific metrics and sample profiler samples, so currently you can't say how long you spend in individual method invocations, but if we did, all timestamps are just performance ticks, so you could, given a thread and method tracing, figure out where time is spend, but I believe it would be more efficient to actually add the start/stop events into our profiler, and once we have them, hook up the EventPipe event to get accurate start/stop durations.